### PR TITLE
Rebuild on page change

### PR DESCRIPTION
**Pull request by wgevaert**
When page owner changes or related subject page changes, rebuild the page;

You want this e.g. when you have [this pull request](https://github.com/WikibaseSolutions/SemanticStructuredDiscussions/pull/13) or if data from [this pull request](https://github.com/WikibaseSolutions/SemanticStructuredDiscussions/pull/14) depends on topic pages.

### CODE REVIEW

File: src/Hooks.php [hunk 1]
```diff
@@ -86,6 +86,24 @@ public static function onBeforeDataUpdateComplete( Store $store, SemanticData $s
 		Services::getDataAnnotator()->addAnnotations( $topic, $semanticData );
 	}
```
Comment (marijnvanwezel) [inline src/Hooks.php:88]: Nitpick: Please add a comment similar to the comment at `onBeforeDataUpdateComplete`.

```suggestion
/**
* Hook to process information after an update has been completed.
*
* @param Store $store
* @param SemanticData $semanticData
* @return void
* @link https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/docs/technical/hooks/hook.store.afterdataupdatecomplete.md
*/
```

File: src/Hooks.php [hunk 2]
```diff
@@ -106,14 +124,29 @@ public static function onAPIFlowAfterExecute( ApiFlowBase $module ): void {
...
 		$dataRebuilder = new DataRebuilder( $store, ServicesFactory::getInstance()->newTitleFactory() );
-		$dataRebuilder->setOptions( new Options( [ 'page' => $page ] ) );
+		$dataRebuilder->setOptions( new Options( [ 'page' => $pageName ] ) );
 		$dataRebuilder->rebuild();
 	}
 
+	/**
+	 * This method is used to reserve the usernames of the system user(s) used by this extension
+	 *
+	 * @link onUserGetReservedNames
```
Comment (marijnvanwezel) [inline src/Hooks.php]: This link links to the method itself.

```suggestion
	 * @link https://www.mediawiki.org/wiki/Manual:Hooks/UserGetReservedNames
```

File: src/SemanticMediaWiki/DataAnnotator.php [hunk 3]
```diff
@@ -51,7 +51,7 @@ public function __construct( AnnotatorStore $annotatorStore ) {
 	 */
 	public function addAnnotations( SDTopic $topic, SemanticData $semanticData ): void {
 		$owner = $topic->getTopicOwner();
-		if ( !$topic->isEveryoneAllowed() || !$owner->exists() || !$owner->getSubjectPage()->exists() ) {
+		if ( !$topic->isEveryoneAllowed() || !$owner || !$owner->exists() || !$owner->getSubjectPage()->exists() ) {
```
Comment (marijnvanwezel) [inline src/SemanticMediaWiki/DataAnnotator.php]: The return type of `getTopicOwner` is not nullable, so `!$owner` will always be false.

File: src/StructuredDiscussions/SDTopic.php [hunk 4]
```diff
@@ -162,6 +162,10 @@ public function getTopicOwner(): Title {
 		return $this->associatedTitle;
 	}
```
Comment (marijnvanwezel) [inline src/StructuredDiscussions/SDTopic.php:164]: Please add a comment explaining this method, and how it differs from `getTopicOwner()`.

File: src/StructuredDiscussions/TopicRepository.php [hunk 5]
```diff
@@ -49,4 +49,28 @@ public function getByTitle( Title $title ): ?SDTopic {
 
 		return new SDTopic( $viewTopic['topic'] );
 	}
+
+	public function getByOwner( Title $title ): array {
+		$parameters = [ 'page' => $title->getTalkPage()->getFullText() ];
```
Comment (marijnvanwezel) [inline src/StructuredDiscussions/TopicRepository.php]: `getTalkPage()` is deprecated.

```suggestion
                $talkPage = $title->getTalkPageIfDefined();
                
                if ( !$talkPage ) {
                    return [];
                }
                
		$parameters = [ 'page' => $talkPage->getFullText() ];
```

File: src/StructuredDiscussions/TopicRepository.php [hunk 6]
```diff
@@ -49,4 +49,28 @@ public function getByTitle( Title $title ): ?SDTopic {
...
+		$topics = $this->callSubmodule( 'view-topiclist', $parameters );
+
+		$rtr = [];
+
+		if ( $topics === null ) {
+			return $rtr;
```
Comment (marijnvanwezel) [inline src/StructuredDiscussions/TopicRepository.php]:
```suggestion
			return [];
```

File: src/StructuredDiscussions/TopicRepository.php [hunk 7]
```diff
@@ -49,4 +49,28 @@ public function getByTitle( Title $title ): ?SDTopic {
...
+		}
+
+		if ( !isset( $topics['topiclist'] ) ) {
+			return $rtr;
```
Comment (marijnvanwezel) [inline src/StructuredDiscussions/TopicRepository.php]:
```suggestion
			return [];
```

File: src/StructuredDiscussions/TopicRepository.php [hunk 8]
```diff
@@ -49,4 +49,28 @@ public function getByTitle( Title $title ): ?SDTopic {
 
 		return new SDTopic( $viewTopic['topic'] );
 	}
+
+	public function getByOwner( Title $title ): array {
+		$parameters = [ 'page' => $title->getTalkPage()->getFullText() ];
+		$topics = $this->callSubmodule( 'view-topiclist', $parameters );
+
+		$rtr = [];
```
Comment (marijnvanwezel) [inline src/StructuredDiscussions/TopicRepository.php]:
```suggestion

```

File: src/StructuredDiscussions/TopicRepository.php [hunk 9]
```diff
@@ -49,4 +49,28 @@ public function getByTitle( Title $title ): ?SDTopic {
...
+		}
+
+		foreach ( $topics[ 'topiclist' ][ 'roots' ] as $root ) {
+			$copy = $topics[ 'topiclist' ];
+			$copy[ 'roots' ] = [ $root ];
+
+			$rtr []= new SDTopic( $copy );
+		}
+
+		return $rtr;
```
Comment (marijnvanwezel) [inline src/StructuredDiscussions/TopicRepository.php:78]:
```suggestion
		return array_map(function ( array $root ) use ( $topics ) {
			$copy = $topics['topiclist'];
			$copy['roots'] = [ $root ];
			
			return new SDTopic( $copy );
		}, $topics['topiclist']['roots'] );
```

**Review by marijnvanwezel — CHANGES_REQUESTED** [review]

File: src/StructuredDiscussions/TopicRepository.php (hunk 9 above)
Comment (wgevaert) [reply src/StructuredDiscussions/TopicRepository.php:78]: What is the added benefit of using the array_map?
All I can see it does is increase memory usage.

Comment (marijnvanwezel) [reply src/StructuredDiscussions/TopicRepository.php:78]: It is functional :)

But there is no real added benefit, so feel free to ignore.

**Review by marijnvanwezel — APPROVED** [review]

**Pull request merged**